File: src/leetcode/easy/ValidParentheses.java
package leetcode.easy;

import java.util.Stack;

public class ValidParentheses {

    public boolean isValid(String s) {
        Stack<Character> stack = new Stack();

        for (char c : s.toCharArray()) {
            if (stack.isEmpty()) stack.push(c);
            else if ((c == ')' && stack.peek() == '(') ||
                    (c == ']' && stack.peek() == '[') ||
                    (c == '}' && stack.peek() == '{')) stack.pop();
            else stack.push(c);
        }

        return stack.isEmpty();
    }
}

File: python-code/neetcode150/valid_parentheses.py
def is_valid(s: str) -> bool:
    pairs: dict[str, str] = {")": "(", "}": "{", "]": "["}

    validation_stack = []

    for character in s:
        if character in pairs:
            if validation_stack and validation_stack[-1] == pairs[character]:
                validation_stack.pop()
            else:
                validation_stack.append(character)
        else:
            validation_stack.append(character)

    return False if validation_stack else True


if __name__ == "__main__":
    print(is_valid("[]"))